### PR TITLE
[Model] Support SigLIP encoder and alternative decoders for LLaVA models

### LINKED ISSUE #7149: [Bug]: The deployment function requires a divisible relationship, but the model structure does not meet this requirement. What should I do?
### 🐛 Describe the bug

def get_siglip_patch_grid_length(*, image_size: int, patch_size: int) -> int:
    assert image_size % patch_size == 0
    return image_size // patch_size
    
<html>
<body>

"vision_config": {
--
  | "hidden_act": "gelu_pytorch_tanh",
  | "hidden_size": 1152,
  | "image_size": 384,
  | "intermediate_size": 4304,
  | "layer_norm_eps": 1e-06,
  | "model_type": "siglip_vision_model",
  | "num_attention_heads": 16,
  | "num_hidden_layers": 27,
  | "patch_size": 14

</body>
</html>

### PR DESCRIPTION
**Pull request by DarkLight1337**
Based on #7067, this PR partially addresses #7143 by enabling SigLIP to be used in LLaVA models. The same code also enables them to load other LLM decoders.

This PR also fixes two hidden bugs in the calculation for the number of placeholder tokens:
- For CLIP encoder, the feature size should be increased by 1 because the CLS token is prepended to the output.
- For LLaVA models, the feature size should be reduced by 1 when `vision_feature_select_strategy="default"`.

These two miscalculations cancelled out each other prior to this PR. They were discovered when trying to support SigLIP in #7149 because SigLIP encoder, unlike CLIP encoder, does not output the CLS token, resulting in an off-by-one error when filling in placeholders.

FIX #7149

### CODE REVIEW

**Comment by BrenchCC** [comment]
I have tried this method 3 hour ago. Bue the error as follos:
[rank0]:   File "/data/vayu/train/workspace/vllm/vllm/model_executor/models/utils.py", line 29, in merge_vision_embeddings
[rank0]:     raise ValueError(
[rank0]: ValueError: Attempted to assign 11 x 728 = 8008 image tokens to 8019 placeholders
It's seems that the model need more image tokens.

**Comment by DarkLight1337** [comment]
> I have tried this method 3 hour ago. Bue the error as follos: [rank0]: File "/data/vayu/train/workspace/vllm/vllm/model_executor/models/utils.py", line 29, in merge_vision_embeddings [rank0]: raise ValueError( [rank0]: ValueError: Attempted to assign 11 x 728 = 8008 image tokens to 8019 placeholders It's seems that the model need more image tokens.

Can you show the input to your model? Also note that we don't support multiple image input per prompt for now.

**Comment by DarkLight1337** [comment]
OK I have found an issue with the image feature size calculation. Fixing...

**Comment by DarkLight1337** [comment]
Fixed, can you try again?

**Comment by BrenchCC** [comment]
I will try it maybe reply tonight or tomorrow.

发自我的iPhone


------------------ Original ------------------
From: Cyrus Leung ***@***.***&gt;
Date: Mon, Aug 5, 2024 6:44 PM
To: vllm-project/vllm ***@***.***&gt;
Cc: Brench ***@***.***&gt;, Comment ***@***.***&gt;
Subject: Re: [vllm-project/vllm] [Model] Support alternative LLM backbone andSigLIP encoder for LLaVA model (PR #7153)





 
Fixed, can you try again?

**Comment by BrenchCC** [comment]
can you try this python file as follows?
'                                                                                                      
import requests
from PIL import Image

from vllm import LLM, SamplingParams


def run_llava_next():
    llm = LLM(model="TIGER-Lab/Mantis-8B-siglip-llama3", max_model_len=4096)

    prompt = " <image>\nWhat is shown in this image? "
    url = "https://h2o-release.s3.amazonaws.com/h2ogpt/bigben.jpg"
    image = Image.open(BytesIO(requests.get(url).content))
    sampling_params = SamplingParams(temperature=0.8,
                                     top_p=0.95,
                                     max_tokens=100)

    outputs = llm.generate(
        {
            "prompt": prompt,
            "multi_modal_data": {
                "image": image
            }
        },
        sampling_params=sampling_params)

    generated_text = ""
    for o in outputs:
        generated_text += o.outputs[0].text

    print(f"LLM output:{generated_text}")


if __name__ == "__main__":
    run_llava_next()
`
because my environment has some problems : '_OpNamespace' '_C' object has no attribute 'rms_norm'

**Comment by BrenchCC** [comment]
`
import requests
from PIL import Image
from vllm import LLM, SamplingParams
def run_llava_next():
    llm = LLM(model="TIGER-Lab/Mantis-8B-siglip-llama3", max_model_len=4096)
    prompt = " <image>\nWhat is shown in this image? "
    url = "https://h2o-release.s3.amazonaws.com/h2ogpt/bigben.jpg"
    image = Image.open(BytesIO(requests.get(url).content))
    sampling_params = SamplingParams(temperature=0.8,
                                     top_p=0.95,
                                     max_tokens=100)
    outputs = llm.generate(
        {
            "prompt": prompt,
            "multi_modal_data": {
                "image": image
            }
        },
        sampling_params=sampling_params)
    generated_text = ""
    for o in outputs:
        generated_text += o.outputs[0].text
    print(f"LLM output:{generated_text}")
if __name__ == "__main__":
    run_llava_next()
`
can you try this? because my conda env has a problem:  
[rank0]: AttributeError: '_OpNamespace' '_C' object has no attribute 'rms_norm'

**Comment by DarkLight1337** [comment]
> from io import BytesIO
> 
> import requests
> from PIL import Image
> 
> from vllm import LLM, SamplingParams
> 
> 
> def run_llava_next():
>     llm = LLM(model="TIGER-Lab/Mantis-8B-siglip-llama3", max_model_len=4096)
> 
>     prompt = " <image>\nWhat is shown in this image? "
>     url = "https://h2o-release.s3.amazonaws.com/h2ogpt/bigben.jpg"
>     image = Image.open(BytesIO(requests.get(url).content))
>     sampling_params = SamplingParams(temperature=0.8,
>                                      top_p=0.95,
>                                      max_tokens=100)
> 
>     outputs = llm.generate(
>         {
>             "prompt": prompt,
>             "multi_modal_data": {
>                 "image": image
>             }
>         },
>         sampling_params=sampling_params)
> 
>     generated_text = ""
>     for o in outputs:
>         generated_text += o.outputs[0].text
> 
>     print(f"LLM output:{generated_text}")
> 
> 
> if __name__ == "__main__":
>     run_llava_next()

You should delete any compiled libraries from your local vLLM repo, and reinstall vLLM.

**Comment by DarkLight1337** [comment]
Based on the test results, the model can run in vLLM but the output is gibberish (probably something wrong with the input processor). Since this PR isn't aimed at fully implementing Mantis, we can create another PR after this one to resolve those issues.

**Comment by BrenchCC** [comment]
I know the mistake. I need to build on CUDA. Your branch is base on AMD. LOL

**Comment by BrenchCC** [comment]
Thanks for your help!!!

**Comment by BrenchCC** [comment]
Yep,because there are some modifications of the 'LlavaForConditionalGeneration' in Mantis about the tokens. So Mantis depends on its own class to use.

File: vllm/model_executor/models/utils.py [hunk 1]
```diff
@@ -1,22 +1,70 @@
...
+        :ref:`init_inner_model`
+    """
+    for name, loaded_weight in weights:
+        name = name.split(".")
+        if prefix == name.pop(0):
+            name = ".".join(name)
+            yield name, loaded_weight
+
+
+def init_inner_model(
```
Comment (ywang96) [inline vllm/model_executor/models/utils.py:31]: Why don't we name this `init_language_model`? I think the word `inner` is pretty confusing for developers. Is there a case where we use this function to initialize a non-LM model?

**Review by ywang96 — APPROVED** [review]
Overall LGTM but I left a few comments - PTAL!

File: tests/models/test_llava.py [hunk 2]
```diff
@@ -29,12 +32,15 @@ def vllm_to_hf_output(vllm_output: Tuple[List[int], str,
     """Sanitize vllm output to be comparable with hf output."""
     output_ids, output_str, out_logprobs = vllm_output
 
+    config = AutoConfig.from_pretrained(model)
+    image_token_id = config.image_token_index
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply tests/models/test_llava.py:36]: I think it is safe to assume this, since `image_token_index` is an attribute in `LlavaConfig`.

File: requirements-test.txt [hunk 3]
```diff
@@ -19,6 +19,7 @@ ray
 sentence-transformers # required for embedding
 compressed-tensors==0.4.0 # required for compressed-tensors
 timm # required for internvl test
+git+https://github.com/TIGER-AI-Lab/Mantis.git # required for llava(mantis) test
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply requirements-test.txt:22]: Sure, I'll comment this out and add a TODO then.

File: tests/models/test_llava.py [hunk 4]
```diff
@@ -67,6 +73,16 @@ def run_test(
     Note, the text input is also adjusted to abide by vllm contract.
     The text output is sanitized to be able to compare with hf.
     """
+    if model.startswith("TIGER-Lab/Mantis"):
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply tests/models/test_llava.py:76]: Done

File: vllm/model_executor/models/utils.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/utils.py:31]: This can be used for any model registered in vLLM (not just LMs), since it shares the same loading logic. The recipe introduced by #7067 can be applied to any inner model that implements weight loading logic.

Comment (ywang96) [reply vllm/model_executor/models/utils.py:31]: I think it's a just a bit confusing when we have a `_init_vision_tower` and `init_inner_model` together, and perhaps we can think of a better name for this pattern other than `inner`. How about explicitly calling it out as `init_registered_model`?

Comment (DarkLight1337) [reply vllm/model_executor/models/utils.py:31]: Hmm... I'm thinking of `init_vllm_model`. Would that work?

Comment (ywang96) [reply vllm/model_executor/models/utils.py:31]: Let's do `init_vllm_registered_model` - I know it's a bit wordy, but I think the purpose of this function will be a lot easier to tell from its name.

**Comment by BrenchCC** [comment]
may be should use some function to remove the unexpected tag.
reference:[https://github.com/TIGER-AI-Lab/Mantis/blob/main/mantis/models/mllava/utils.py](url)
```
import PIL
import torch
from .modeling_llava import LlavaForConditionalGeneration
from .processing_llava import MLlavaProcessor
# from ..conversation import conv_mllava_v1_mmtag as default_conv
from ..conversation import conv_mllava_v1 as default_conv, conv_templates

from typing import List, Tuple, Union, Tuple

def chat_mllava(
    text:str, 
    images: List[Union[PIL.Image.Image, str]], 
    model:LlavaForConditionalGeneration, 
    processor:MLlavaProcessor, 
    max_input_length:int=None, 
    history:List[dict]=None, 
    **kwargs) -> Tuple[str, List[dict]]:
    """
    Chat with the Mllava model
    Args:
        text: str, the text to be sent to the model, where <image> will be the placeholder for the image
        images: List[PIL.Image.Image], the images to be sent to the model, or None  
        model: LlavaForConditionalGeneration, the model to be used
        processor: MLlavaProcessor, the processor to be used
        max_input_length: int, the maximum input length
        history: List[dict], list of messages in the conversation as history. Each message is a dictionary {"role": "ASSISTANT/USER", "text": "the message"}. If None, the conversation will start from scratch
        kwargs: dict, the generation kwargs
    Returns:
        Tuple[str, List[dict]], the generated text and the history of the conversation
        

    """
    if "llama-3" in model.language_model.name_or_path.lower():
        conv = conv_templates['llama_3']
        terminators = [
            processor.tokenizer.eos_token_id,
            processor.tokenizer.convert_tokens_to_ids("<|eot_id|>")
        ]
    else:
        conv = default_conv
        terminators = None
    kwargs["eos_token_id"] = terminators
    conv = conv.copy()
    conv.messages = []
    if history is not None:
        for message in history:
            assert message["role"] in conv.roles
            conv.append_message(message["role"], message["text"])
        if text:
            assert conv.messages[-1][0] == conv.roles[1], "The last message in the history should be the assistant, if the given text is not empty"
            conv.append_message(conv.roles[0], text)
            conv.append_message(conv.roles[1], "")
            history.append({"role": conv.roles[0], "text": text})
            history.append({"role": conv.roles[1], "text": ""})
        else:
            if conv.messages[-1][0] == conv.roles[1]:
                assert conv.messages[-1][1] == "", "No user message should be provided"
            else:
                assert conv.messages[-1][0] == conv.roles[0], "The last message in the history should be the user, if the given text is empty"
                conv.append_message(conv.roles[0], "")
                history.append({"role": conv.roles[0], "text": ""})
    else:
        history = []
        history.append({"role": conv.roles[0], "text": text})
        history.append({"role": conv.roles[1], "text": ""})
        conv.append_message(conv.roles[0], text)
        conv.append_message(conv.roles[1], "")
    assert conv.messages[-1][0] == conv.roles[1] and conv.messages[-1][1] == "", "Format check"
    assert history[-1]["role"] == conv.roles[1] and history[-1]["text"] == "", "Format check"
    
    prompt = conv.get_prompt()
    if images:
        for i in range(len(images)):
            if isinstance(images[i], str):
                images[i] = PIL.Image.open(images[i]).convert("RGB")
    
    inputs = processor(images=images, text=prompt, return_tensors="pt", truncation=True, max_length=max_input_length)
    for k, v in inputs.items():
        if v is not None:
            if isinstance(v, torch.Tensor):
                inputs[k] = v.to(model.device)
            elif isinstance(v, list):
                inputs[k] = [x.to(model.device) for x in v]
            else:
                raise ValueError(f"Invalid input type: {type(v)}")
    

    output_ids = model.generate(**inputs, **kwargs)
    output_ids = output_ids[0]
    
    # remove the input tokens
    generated_ids = output_ids[inputs["input_ids"].shape[-1]:]
    generated_text = processor.decode(generated_ids, skip_special_tokens=True)

    history[-1]["text"] = generated_text
    
    return generated_text, history
```
[reference:](https://github.com/open-compass/VLMEvalKit/blob/main/vlmeval/vlm/mantis.py)
```
import torch
from PIL import Image
from abc import abstractproperty
from .base import BaseModel
from ..smp import *
from ..dataset import DATASET_TYPE
import warnings


class Mantis(BaseModel):
    """
    Mantis Model
    This implementation is adpated from the Llava model from llava.py and the Idefics model from idefics.py
    """
    INSTALL_REQ = True
    INTERLEAVE = True

    DEFAULT_IMAGE_TOKEN = '<image>'
    IMAGE_TOKEN_INDEX = -200

    def __init__(self, model_path='TIGER-Lab/Mantis-8B-siglip-llama3', **kwargs):
        assert model_path is not None
        try:
            from mantis.models.mllava import LlavaForConditionalGeneration, MLlavaProcessor
            from mantis.models.mfuyu import MFuyuForCausalLM, MFuyuProcessor
            from mantis.models.conversation import conv_mllava_v1 as default_conv, conv_templates
        except:
            warnings.warn(
                "Mantis is not installed. Please install Mantis to use this model.Please use 'pip install "
                "git+https://github.com/TIGER-AI-Lab/Mantis.git' to install"
            )

        try:
            from transformers import AutoModelForVision2Seq, AutoProcessor
        except Exception as e:
            warnings.warn("Upgrade transformers to use Mantis's idefics model.\nError: %s" % e)
        except:
            warnings.warn('Please `pip install git+https://github.com/LLaVA-VL/LLaVA-NeXT.git')

        # inference implementation for attention, can be "sdpa", "eager", "flash_attention_2".
        # Seems FA2 is not effective during inference:
        # https://discuss.huggingface.co/t/flash-attention-has-no-effect-on-inference/73453/5
        # if is_flash_attn_2_available:
        #     best_fit_attn_implementation = "flash_attention_2"
        # flash_attn has a bug that says: ERROR Error query and key must have the same dtype in generating

        try:
            import flash_attn
            best_fit_attn_implementation = 'flash_attention_2'
        except ImportError:
            best_fit_attn_implementation = 'eager'
        self.model_path = model_path
        attn_implementation = best_fit_attn_implementation
        self._is_idefics = 'idefics' in model_path.lower()
        # Here load the "non-idefics" Mantis model.
        if not self._is_idefics:
            if 'fuyu' in model_path.lower():
                self.processor = MFuyuProcessor.from_pretrained(self.model_path)
                model = MFuyuForCausalLM.from_pretrained(
                    self.model_path,
                    device_map='cuda',
                    attn_implementation=attn_implementation,
                    torch_dtype=torch.float16
                )
            else:
                self.processor = MLlavaProcessor.from_pretrained(self.model_path)
                model = LlavaForConditionalGeneration.from_pretrained(
                    self.model_path,
                    device_map='cuda',
                    attn_implementation=attn_implementation,
                    torch_dtype=torch.float16
                )
        else:
            self.processor = AutoProcessor.from_pretrained(self.model_path)
            model = AutoModelForVision2Seq.from_pretrained(
                self.model_path,
                device_map='cuda',
                torch_dtype=torch.float16
            )

        model = model.eval()
        self.model = model.cuda()
        kwargs_default = dict(do_sample=False, temperature=0, max_new_tokens=1024, top_p=None, num_beams=1)
        kwargs_default.update(kwargs)
        self.kwargs = kwargs_default
        warnings.warn(f'Following kwargs received: {self.kwargs}, will use as generation config. ')

        self.tokenizer = self.processor.tokenizer
        self.default_conv = default_conv
        self.conv_templates = conv_templates

    def use_custom_prompt(self, dataset):
        assert dataset is not None
        if DATASET_TYPE(dataset) == 'MCQ':
            return True
        return False

    def build_prompt(self, line, dataset=None):
        assert self.use_custom_prompt(dataset)
        assert dataset is None or isinstance(dataset, str)
        tgt_path = self.dump_image(line, dataset)

        question = line['question']
        hint = line['hint'] if ('hint' in line and not pd.isna(line['hint'])) else None
        if hint is not None:
            question = hint + '\n' + question
        options = {
            cand: line[cand]
            for cand in string.ascii_uppercase
            if cand in line and not pd.isna(line[cand])
        }
        for key, item in options.items():
            question += f'\n{key}. {item}'
        prompt = question

        if len(options):
            prompt += (
                '\n请直接回答选项字母。' if cn_string(prompt) else
                "\nAnswer with the option's letter from the given choices directly."
            )
        else:
            prompt += '\n请直接回答问题。' if cn_string(prompt) else '\nAnswer the question directly.'
        message = [dict(type='image', value=s) for s in tgt_path]
        message.append(dict(type='text', value=prompt))
        return message

    def output_process(self, answer):
        if '<s>' in answer:
            answer = answer.replace('<s>', '').strip()
        if '[/INST]' in answer:
            answer = answer.split('[/INST]')[1].strip()
        elif 'ASSISTANT:' in answer:
            answer = answer.split('ASSISTANT:')[1].strip()
        elif 'assistant\n' in answer:
            answer = answer.split('assistant\n')[1].strip()
        elif '<|end_header_id|>\n\n' in answer:
            answer = answer.split('<|end_header_id|>\n\n')[2].strip()

        if '</s>' in answer:
            answer = answer.split('</s>')[0].strip()
        elif '<|im_end|>' in answer:
            answer = answer.split('<|im_end|>')[0].strip()
        elif '<|eot_id|>' in answer:
            answer = answer.split('<|eot_id|>')[0].strip()
        elif '<end_of_utterance>':
            answer = answer.split('<end_of_utterance>')[0].strip()
        return answer

    def generate_inner(self, message, dataset=None):
        content, images = '', []
        ide_content, question = [], ''
        for msg in message:
            if msg['type'] == 'text':
                content += msg['value']
                question += msg['value']
            else:
                images.append(Image.open(msg['value']).convert('RGB'))
                content += (self.DEFAULT_IMAGE_TOKEN + '\n')
                ide_content.append({'type': 'image'})
        if self._is_idefics:
            # Follow the idefics implementation:
            ide_content.append({'type': 'text', 'text': question})
            prompt = [{'role': 'user', 'content': ide_content}]
            prompt = self.processor.apply_chat_template(prompt, add_generation_prompt=True)
        else:
            # Follow the Mantis code base to make sure they are consistent:
            # https://github.com/TIGER-AI-Lab/Mantis/blob/main/mantis/models/mllava/utils.py#L33
            # Users don't need to define chat template as it is done here
            if 'llama-3' in self.model.language_model.name_or_path.lower():
                conv = self.conv_templates['llama_3']
                terminators = [
                    self.processor.tokenizer.eos_token_id,
                    self.processor.tokenizer.convert_tokens_to_ids('<|eot_id|>')
                ]
            else:
                conv = self.default_conv
                terminators = None

            # Using EOT because end of *text* is more accurate for what we're doing than end of *sentence*
            if 'eos_token_id' not in self.kwargs:
                self.kwargs['eos_token_id'] = terminators

            conv = conv.copy()
            conv.append_message(conv.roles[0], content)
            conv.append_message(conv.roles[1], '')
            assert conv.messages[-1][0] == conv.roles[1] and conv.messages[-1][1] == '', 'Format check'
            prompt = conv.get_prompt()

        inputs = self.processor(prompt, images, return_tensors='pt', truncation=True)
        # FIXME: Fuyu model would return a list instead of a pytorch tensor. This weird behavior needs fixing.
        if 'image_patches' in inputs.keys():
            inputs['image_patches'] = inputs['image_patches'][0]
        inputs = {k: v.to(self.model.device) for k, v in inputs.items()}
        output = self.model.generate(**inputs, **self.kwargs)
        output = output[0]
        generated_ids = output[inputs['input_ids'].shape[-1]:]
        answer = self.processor.decode(generated_ids, skip_special_token=True)
        answer = self.output_process(answer)
        return answer
```
